In [2]:
# we import the modules/libraries that we'll need
import pandas as pd

### Importing Data, Diagnostics, and Data Cleaning

For this tutorial, we're going to use the "Correlates of State Policy Project" dataset, from the IPPSR of Michigan State University, which can be accessed at the following [link](https://ippsr.msu.edu/public-policy/correlates-state-policy).


In order to import our data, we're going to use the `pandas` library, which we've imported above. In particular, we're going to use the function `read_csv()`, as our sample dataset is in csv format. There are analogous functions for other kinds of formats, such as `read_excel` for Excel file types, such as `xls`, and `xlsx`.

In [3]:
# How to import our data
# For pedagogical purposes, we'll write explicitly the key parameters we're going to need 
df = pd.read_csv(filepath_or_buffer= "msu_dataset.csv", 
                 sep = ",",
                 index_col= 0,
                 parse_dates= True,
                 float_precision= "round_trip")

In [ ]:
# This simple command allows us to display the first 5 (or more) rows of our dataframe
# It's always good to preliminarily have a quick look to make sure the import was successful
df.head()

st  stateno   state  state_fips  state_icpsr  poptotal  \
year                                                                 
1900-01-01  AK      2.0  Alaska           2           81       NaN   
1901-01-01  AK      2.0  Alaska           2           81       NaN   
1902-01-01  AK      2.0  Alaska           2           81       NaN   
1903-01-01  AK      2.0  Alaska           2           81       NaN   
1904-01-01  AK      2.0  Alaska           2           81       NaN   

            evangelical_pop  nonwhite  gini_coef  region  ...  solaw  vaaban  \
year                                                      ...                  
1900-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   
1901-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   
1902-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   
1903-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   
1904-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   

            w_environment_solar_taxcredit  frps  drugs_medical_marijuana  \
year                                                                       
1900-01-01                            NaN   NaN                      NaN   
1901-01-01                            NaN   NaN                      NaN   
1902-01-01                            NaN   NaN                      NaN   
1903-01-01                            NaN   NaN                      NaN   
1904-01-01                            NaN   NaN                      NaN   

            gambling_lottery_adoption  guncontrol_assaultweapon_ban  \
year                                                                  
1900-01-01                        NaN                           NaN   
1901-01-01                        NaN                           NaN   
1902-01-01                        NaN                           NaN   
1903-01-01                        NaN                           NaN   
1904-01-01                        NaN                           NaN   

            guncontrol_opencarry  labor_minwage_abovefed  labor_right_to_work  
year                                                                           
1900-01-01                   NaN                     NaN                  NaN  
1901-01-01                   NaN                     NaN                  NaN  
1902-01-01                   NaN                     NaN                  NaN  
1903-01-01                   NaN                     NaN                  NaN  
1904-01-01                   NaN                     NaN                  NaN  

[5 rows x 71 columns]

In [6]:
# The following command allows us to have a look at the shape of our dataset
# In this case, as the table is 2-dimensional, it'll only display rows and columns
# In other cases, it will show the additional dimensions, for more complex data structures

df.shape

(6171, 71)

Here above we can see that our dataset has 6171 observations/rows, and 71 columns. 

In [8]:
# This is a quick command to check the NaN observations
df.isna()

st  stateno  state  state_fips  state_icpsr  poptotal  \
year                                                                   
1900-01-01  False    False  False       False        False      True   
1901-01-01  False    False  False       False        False      True   
1902-01-01  False    False  False       False        False      True   
1903-01-01  False    False  False       False        False      True   
1904-01-01  False    False  False       False        False      True   
...           ...      ...    ...         ...          ...       ...   
2016-01-01  False    False  False       False        False     False   
2017-01-01  False    False  False       False        False     False   
2018-01-01  False    False  False       False        False     False   
2019-01-01  False    False  False       False        False     False   
2020-01-01  False    False  False       False        False      True   

            evangelical_pop  nonwhite  gini_coef  region  ...  solaw  vaaban  \
year                                                      ...                  
1900-01-01             True      True       True    True  ...   True    True   
1901-01-01             True      True       True    True  ...   True    True   
1902-01-01             True      True       True    True  ...   True    True   
1903-01-01             True      True       True    True  ...   True    True   
1904-01-01             True      True       True    True  ...   True    True   
...                     ...       ...        ...     ...  ...    ...     ...   
2016-01-01             True      True       True   False  ...   True   False   
2017-01-01             True      True       True    True  ...   True    True   
2018-01-01             True      True       True    True  ...   True    True   
2019-01-01             True      True       True    True  ...   True    True   
2020-01-01             True      True       True    True  ...   True    True   

            w_environment_solar_taxcredit   frps  drugs_medical_marijuana  \
year                                                                        
1900-01-01                           True   True                     True   
1901-01-01                           True   True                     True   
1902-01-01                           True   True                     True   
1903-01-01                           True   True                     True   
1904-01-01                           True   True                     True   
...                                   ...    ...                      ...   
2016-01-01                           True  False                     True   
2017-01-01                           True   True                     True   
2018-01-01                           True   True                     True   
2019-01-01                           True   True                     True   
2020-01-01                           True   True                     True   

            gambling_lottery_adoption  guncontrol_assaultweapon_ban  \
year                                                                  
1900-01-01                       True                          True   
1901-01-01                       True                          True   
1902-01-01                       True                          True   
1903-01-01                       True                          True   
1904-01-01                       True                          True   
...                               ...                           ...   
2016-01-01                       True                          True   
2017-01-01                       True                          True   
2018-01-01                       True                          True   
2019-01-01                       True                          True   
2020-01-01                       True                          True   

            guncontrol_opencarry  labor_minwage_abovefed  labor_right_to_work  
year                           

In [13]:
# To summarize the number of NaN values of each column, we can use the following command
df.isna().sum()

st                                 0
stateno                            0
state                              0
state_fips                         0
state_icpsr                        0
                                ... 
gambling_lottery_adoption       3570
guncontrol_assaultweapon_ban    4845
guncontrol_opencarry            3417
labor_minwage_abovefed          3921
labor_right_to_work             2550
Length: 71, dtype: int64

In [14]:
# To look more in detail at the number of NaN values that a specific column has, first we may want to have
# a look at the names of all the columns

df.columns

Index(['st', 'stateno', 'state', 'state_fips', 'state_icpsr', 'poptotal',
       'evangelical_pop', 'nonwhite', 'gini_coef', 'region',
       'total_expenditure', 'total_revenue', 'gsppcap', 'incomepcap',
       'unemployment', 'povrate', 'earned_income_taxcredit', 'x_sales_taxes',
       'x_tax_burden', 'x_top_corporateincometaxrate', 'prez_election_year',
       'hou_chamber', 'sen_chamber', 'h_diffs', 's_diffs', 'inst6014_nom',
       'bowen_legprof_firstdim', 'bowen_legprof_seconddim', 'speaker_power',
       'pctfemaleleg', 'efna_index', 'vepvotingrate', 'ranney4_control',
       'overall_fin_reg', 'oelecsc', 'opartsc', 'nonindiv_contrib',
       'union_density', 'pollib_median', 'innovatescore_boehmkeskinner',
       'soc_capital_ma', 'citi6013', 'wpid', 'mood', 'anti_environment',
       'anti_aid', 'anti_defense', 'anti_education', 'anti_welfare',
       'anti_race', 'death_penalty', 'guncontrol_stand_your_ground',
       'vcrimerate', 'z_education_expenditures_per_pup', 'maths

In [16]:
# to get more detailed diagnostics, we can perform the following
# example with guncontrol_assaultweappon_ban
# by inserting the string inside square brackets, we index, so we focus on a specific column
# and then we can perform manipulations on it

df["guncontrol_assaultweapon_ban"].isna().sum()

4845

In [27]:
# if you want to drop all the rows containing a NaN value, you can perform the following
# axis=0 indicates that we're working with respect to rows

# the threshold parameter allows you to set a minimum number of non-NaN values needed not to drop  
df_dna_row = df.dropna(axis = 0, thresh= 100)
# with this we check the shape of the resulting table
df_dna_row.shape

(0, 71)

Here we can see that, as all rows have at least a NaN value, by dropping all such rows, we get a table with 0 rows, essentially.

In [ ]:
# now, let's do the same with respect to columns
# to do so, we must set axis=1

# the how parameter with "any" drops a column if any NaN/NA values are present
# the how parameter with "all" drops a column if any NaN/NA values are present
df_dna_col = df.dropna(axis = 1, how="any")
df_dna_col.shape

(6171, 5)

Here we can see that all but 5 columns have at least a NaN value, and therefore they've been dropped.
Let's have a look at what these columns are.

The `isnull()` command is the same, but for the case of `None` values instead of `NaN` values.

A way to manage missing values is to use the function `fillna()`. 

In [30]:
# A basic, but not always optimal solution, may be to assign the value 0 to all missing values
# Reminder: we use head to only display the first 5 rows
df.fillna(value = 0).head()

st  stateno   state  state_fips  state_icpsr  poptotal  \
year                                                                 
1900-01-01  AK      2.0  Alaska           2           81       0.0   
1901-01-01  AK      2.0  Alaska           2           81       0.0   
1902-01-01  AK      2.0  Alaska           2           81       0.0   
1903-01-01  AK      2.0  Alaska           2           81       0.0   
1904-01-01  AK      2.0  Alaska           2           81       0.0   

            evangelical_pop  nonwhite  gini_coef  region  ...  solaw  vaaban  \
year                                                      ...                  
1900-01-01              0.0       0.0        0.0     0.0  ...    0.0     0.0   
1901-01-01              0.0       0.0        0.0     0.0  ...    0.0     0.0   
1902-01-01              0.0       0.0        0.0     0.0  ...    0.0     0.0   
1903-01-01              0.0       0.0        0.0     0.0  ...    0.0     0.0   
1904-01-01              0.0       0.0        0.0     0.0  ...    0.0     0.0   

            w_environment_solar_taxcredit  frps  drugs_medical_marijuana  \
year                                                                       
1900-01-01                            0.0   0.0                      0.0   
1901-01-01                            0.0   0.0                      0.0   
1902-01-01                            0.0   0.0                      0.0   
1903-01-01                            0.0   0.0                      0.0   
1904-01-01                            0.0   0.0                      0.0   

            gambling_lottery_adoption  guncontrol_assaultweapon_ban  \
year                                                                  
1900-01-01                        0.0                           0.0   
1901-01-01                        0.0                           0.0   
1902-01-01                        0.0                           0.0   
1903-01-01                        0.0                           0.0   
1904-01-01                        0.0                           0.0   

            guncontrol_opencarry  labor_minwage_abovefed  labor_right_to_work  
year                                                                           
1900-01-01                   0.0                     0.0                  0.0  
1901-01-01                   0.0                     0.0                  0.0  
1902-01-01                   0.0                     0.0                  0.0  
1903-01-01                   0.0                     0.0                  0.0  
1904-01-01                   0.0                     0.0                  0.0  

[5 rows x 71 columns]

In [34]:
# We can fill the holes in our reindexed Series by using "ffil" or "bfill"
# "ffil" propagates the last valid observation forward to the next vali
# "bfill" propagates the next valid observation to fill the gaps

df.fillna(method = "bfill").head()

st  stateno   state  state_fips  state_icpsr  poptotal  \
year                                                                 
1900-01-01  AK      2.0  Alaska           2           81  135000.0   
1901-01-01  AK      2.0  Alaska           2           81  135000.0   
1902-01-01  AK      2.0  Alaska           2           81  135000.0   
1903-01-01  AK      2.0  Alaska           2           81  135000.0   
1904-01-01  AK      2.0  Alaska           2           81  135000.0   

            evangelical_pop  nonwhite  gini_coef  region  ...  solaw  vaaban  \
year                                                      ...                  
1900-01-01             11.9  0.205493   0.443897     2.0  ...    0.0     0.0   
1901-01-01             11.9  0.205493   0.443897     2.0  ...    0.0     0.0   
1902-01-01             11.9  0.205493   0.443897     2.0  ...    0.0     0.0   
1903-01-01             11.9  0.205493   0.443897     2.0  ...    0.0     0.0   
1904-01-01             11.9  0.205493   0.443897     2.0  ...    0.0     0.0   

            w_environment_solar_taxcredit  frps  drugs_medical_marijuana  \
year                                                                       
1900-01-01                            0.0   0.0                      0.0   
1901-01-01                            0.0   0.0                      0.0   
1902-01-01                            0.0   0.0                      0.0   
1903-01-01                            0.0   0.0                      0.0   
1904-01-01                            0.0   0.0                      0.0   

            gambling_lottery_adoption  guncontrol_assaultweapon_ban  \
year                                                                  
1900-01-01                        0.0                           0.0   
1901-01-01                        0.0                           0.0   
1902-01-01                        0.0                           0.0   
1903-01-01                        0.0                           0.0   
1904-01-01                        0.0                           0.0   

            guncontrol_opencarry  labor_minwage_abovefed  labor_right_to_work  
year                                                                           
1900-01-01                   0.0                     1.0                  0.0  
1901-01-01                   0.0                     1.0                  0.0  
1902-01-01                   0.0                     1.0                  0.0  
1903-01-01                   0.0                     1.0                  0.0  
1904-01-01                   0.0                     1.0                  0.0  

[5 rows x 71 columns]

In [36]:
df.fillna(method = "ffill").head()

st  stateno   state  state_fips  state_icpsr  poptotal  \
year                                                                 
1900-01-01  AK      2.0  Alaska           2           81       NaN   
1901-01-01  AK      2.0  Alaska           2           81       NaN   
1902-01-01  AK      2.0  Alaska           2           81       NaN   
1903-01-01  AK      2.0  Alaska           2           81       NaN   
1904-01-01  AK      2.0  Alaska           2           81       NaN   

            evangelical_pop  nonwhite  gini_coef  region  ...  solaw  vaaban  \
year                                                      ...                  
1900-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   
1901-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   
1902-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   
1903-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   
1904-01-01              NaN       NaN        NaN     NaN  ...    NaN     NaN   

            w_environment_solar_taxcredit  frps  drugs_medical_marijuana  \
year                                                                       
1900-01-01                            NaN   NaN                      NaN   
1901-01-01                            NaN   NaN                      NaN   
1902-01-01                            NaN   NaN                      NaN   
1903-01-01                            NaN   NaN                      NaN   
1904-01-01                            NaN   NaN                      NaN   

            gambling_lottery_adoption  guncontrol_assaultweapon_ban  \
year                                                                  
1900-01-01                        NaN                           NaN   
1901-01-01                        NaN                           NaN   
1902-01-01                        NaN                           NaN   
1903-01-01                        NaN                           NaN   
1904-01-01                        NaN                           NaN   

            guncontrol_opencarry  labor_minwage_abovefed  labor_right_to_work  
year                                                                           
1900-01-01                   NaN                     NaN                  NaN  
1901-01-01                   NaN                     NaN                  NaN  
1902-01-01                   NaN                     NaN                  NaN  
1903-01-01                   NaN                     NaN                  NaN  
1904-01-01                   NaN                     NaN                  NaN  

[5 rows x 71 columns]

Alternatively, we may use more a slightly more sophisticated method where we use a dictionary to choose for each column we're interested a particular fill value.

In [41]:
values = {"poptotal":10, "evangelical_pop": 5}

df.fillna(value = values).head()

st  stateno   state  state_fips  state_icpsr  poptotal  \
year                                                                 
1900-01-01  AK      2.0  Alaska           2           81      10.0   
1901-01-01  AK      2.0  Alaska           2           81      10.0   
1902-01-01  AK      2.0  Alaska           2           81      10.0   
1903-01-01  AK      2.0  Alaska           2           81      10.0   
1904-01-01  AK      2.0  Alaska           2           81      10.0   

            evangelical_pop  nonwhite  gini_coef  region  ...  solaw  vaaban  \
year                                                      ...                  
1900-01-01              5.0       NaN        NaN     NaN  ...    NaN     NaN   
1901-01-01              5.0       NaN        NaN     NaN  ...    NaN     NaN   
1902-01-01              5.0       NaN        NaN     NaN  ...    NaN     NaN   
1903-01-01              5.0       NaN        NaN     NaN  ...    NaN     NaN   
1904-01-01              5.0       NaN        NaN     NaN  ...    NaN     NaN   

            w_environment_solar_taxcredit  frps  drugs_medical_marijuana  \
year                                                                       
1900-01-01                            NaN   NaN                      NaN   
1901-01-01                            NaN   NaN                      NaN   
1902-01-01                            NaN   NaN                      NaN   
1903-01-01                            NaN   NaN                      NaN   
1904-01-01                            NaN   NaN                      NaN   

            gambling_lottery_adoption  guncontrol_assaultweapon_ban  \
year                                                                  
1900-01-01                        NaN                           NaN   
1901-01-01                        NaN                           NaN   
1902-01-01                        NaN                           NaN   
1903-01-01                        NaN                           NaN   
1904-01-01                        NaN                           NaN   

            guncontrol_opencarry  labor_minwage_abovefed  labor_right_to_work  
year                                                                           
1900-01-01                   NaN                     NaN                  NaN  
1901-01-01                   NaN                     NaN                  NaN  
1902-01-01                   NaN                     NaN                  NaN  
1903-01-01                   NaN                     NaN                  NaN  
1904-01-01                   NaN                     NaN                  NaN  

[5 rows x 71 columns]

If we want to check if a row has a duplicated, then, we can use the `duplicated` method, which will return us a "boolean mask" indicating duplicate rows. In order words, it returns "True" if a row has a duplicate, and "False" otherwise.

In [42]:
df.duplicated()

year
1900-01-01    False
1901-01-01     True
1902-01-01     True
1903-01-01     True
1904-01-01     True
              ...  
2016-01-01    False
2017-01-01    False
2018-01-01    False
2019-01-01    False
2020-01-01    False
Length: 6171, dtype: bool

In [48]:
# in order to eliminate the duplicated rows, we can use the following function "drop_duplicates()"

df_dd = df.drop_duplicates(keep = "first", subset = df.columns)
(df.shape, df_dd.shape)

((6171, 71), (6090, 71))

In [47]:
print(f"We can see that the original dataset had {df.shape[0]-df_dd.shape[0]} duplicated rows")

We can see that the original dataset had 81 duplicated rows


Sometimes the `read_csv()` may fail to recognize the data type of a certain column. In such case, we may want to perform a manual conversion.

In [54]:
# An example of conversion to the type "int", which corresponds to natural numbers is the following
# tail is the analogous of .head(), but it displays the final 5 rows
df["poptotal"].tail()

year
2016-01-01    585501.0
2017-01-01    579315.0
2018-01-01    577601.0
2019-01-01    578759.0
2020-01-01         NaN
Name: poptotal, dtype: float64

We can see that in our case `poptotal` has the float64 datat type, which corresponds to decimal numbers. 
Let's try to convert it to a string format.

In [61]:
df["poptotal"].astype(str)

year
1900-01-01         nan
1901-01-01         nan
1902-01-01         nan
1903-01-01         nan
1904-01-01         nan
                ...   
2016-01-01    585501.0
2017-01-01    579315.0
2018-01-01    577601.0
2019-01-01    578759.0
2020-01-01         nan
Name: poptotal, Length: 6171, dtype: object

In [66]:
# This is how we can refer to the index of our dataframe
df.index

DatetimeIndex(['1900-01-01', '1901-01-01', '1902-01-01', '1903-01-01',
               '1904-01-01', '1905-01-01', '1906-01-01', '1907-01-01',
               '1908-01-01', '1909-01-01',
               ...
               '2011-01-01', '2012-01-01', '2013-01-01', '2014-01-01',
               '2015-01-01', '2016-01-01', '2017-01-01', '2018-01-01',
               '2019-01-01', '2020-01-01'],
              dtype='datetime64[ns]', name='year', length=6171, freq=None)

In [72]:
# To reformat our index to hold show the corresponding year, we can use the strftime command
df.index = df.index.strftime("%Y")

In [73]:
df.index

Index(['1900', '1901', '1902', '1903', '1904', '1905', '1906', '1907', '1908',
       '1909',
       ...
       '2011', '2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019',
       '2020'],
      dtype='object', name='year', length=6171)

We can use indexing to conditionally select only the rows which fulfill a particular condition in the following way

In [79]:
# First we need to reconvert "poptotal" to a numerical format
df["poptotal"] = df["poptotal"].astype(float)

# In the following way we can isolate only the rows where "poptotal" is greater than 500k
df[df["poptotal"]> 500000]

st  stateno    state  state_fips  state_icpsr  poptotal  \
year                                                            
1984  AK      2.0   Alaska           2           81  513702.0   
1985  AK      2.0   Alaska           2           81  532495.0   
1986  AK      2.0   Alaska           2           81  544268.0   
1987  AK      2.0   Alaska           2           81  539309.0   
1988  AK      2.0   Alaska           2           81  541983.0   
...   ..      ...      ...         ...          ...       ...   
2015  WY     50.0  Wyoming          56           68  586107.0   
2016  WY     50.0  Wyoming          56           68  585501.0   
2017  WY     50.0  Wyoming          56           68  579315.0   
2018  WY     50.0  Wyoming          56           68  577601.0   
2019  WY     50.0  Wyoming          56           68  578759.0   

      evangelical_pop  nonwhite  gini_coef  region  ...  solaw  vaaban  \
year                                                ...                  
1984        11.900000  0.235833   0.502557     2.0  ...    0.0     NaN   
1985        16.799999  0.236023   0.504690     2.0  ...    0.0     NaN   
1986        16.799999  0.239796   0.573903     2.0  ...    0.0     NaN   
1987        16.799999  0.245473   0.644675     2.0  ...    0.0     NaN   
1988        16.799999  0.252325   0.695108     2.0  ...    0.0     NaN   
...               ...       ...        ...     ...  ...    ...     ...   
2015              NaN       NaN        NaN     2.0  ...    NaN     0.0   
2016              NaN       NaN        NaN     2.0  ...    NaN     0.0   
2017              NaN       NaN        NaN     NaN  ...    NaN     NaN   
2018              NaN       NaN        NaN     NaN  ...    NaN     NaN   
2019              NaN       NaN        NaN     NaN  ...    NaN     NaN   

      w_environment_solar_taxcredit  frps  drugs_medical_marijuana  \
year                                                                 
1984                            0.0   NaN                      NaN   
1985                            0.0   NaN                      NaN   
1986                            0.0   NaN                      NaN   
1987                            0.0   NaN                      NaN   
1988                            0.0   NaN                      NaN   
...                             ...   ...                      ...   
2015                            NaN   0.0                      NaN   
2016                            NaN   0.0                      NaN   
2017                            NaN   NaN                      NaN   
2018                            NaN   NaN                      NaN   
2019                            NaN   NaN                      NaN   

      gambling_lottery_adoption  guncontrol_assaultweapon_ban  \
year                                                            
1984                        0.0                           NaN   
1985                        0.0                           NaN   
1986                        0.0                           NaN   
1987                        0.0                           NaN   
1988                        0.0                           NaN   
...                         ...                           ...   
2015                        NaN                           NaN   
2016                        NaN                           NaN   
2017                        NaN                           NaN   
2018                        NaN                           NaN   
2019                        NaN                           NaN   

      guncontrol_opencarry  labor_minwage_abovefed  labor_right_to_work  
year                                                                     
1984                   0.0                     1.0                  0.0  
1985                   0.0                     1.0                  0.0  
1986                   0.0                     1.0                  0.0  
1987                   0.0                     1.0                  0.

In [81]:
# To conditionally index columns under multiple conditions, we can perform the following
# We put in curly brackets each condition and use the operators & for "and" and | for "or"

df[(df["state"] == "Alaska")&(df["poptotal"]>532495)].head()

st  stateno   state  state_fips  state_icpsr  poptotal  evangelical_pop  \
year                                                                            
1986  AK      2.0  Alaska           2           81  544268.0        16.799999   
1987  AK      2.0  Alaska           2           81  539309.0        16.799999   
1988  AK      2.0  Alaska           2           81  541983.0        16.799999   
1989  AK      2.0  Alaska           2           81  547159.0        16.799999   
1990  AK      2.0  Alaska           2           81  550043.0        16.799999   

      nonwhite  gini_coef  region  ...  solaw  vaaban  \
year                               ...                  
1986  0.239796   0.573903     2.0  ...    0.0     NaN   
1987  0.245473   0.644675     2.0  ...    0.0     NaN   
1988  0.252325   0.695108     2.0  ...    0.0     NaN   
1989  0.257314   0.667497     2.0  ...    0.0     NaN   
1990  0.259947   0.670734     2.0  ...    0.0     NaN   

      w_environment_solar_taxcredit  frps  drugs_medical_marijuana  \
year                                                                 
1986                            0.0   NaN                      NaN   
1987                            0.0   NaN                      NaN   
1988                            0.0   NaN                      NaN   
1989                            0.0   NaN                      NaN   
1990                            0.0   NaN                      NaN   

      gambling_lottery_adoption  guncontrol_assaultweapon_ban  \
year                                                            
1986                        0.0                           NaN   
1987                        0.0                           NaN   
1988                        0.0                           NaN   
1989                        0.0                           0.0   
1990                        0.0                           0.0   

      guncontrol_opencarry  labor_minwage_abovefed  labor_right_to_work  
year                                                                     
1986                   0.0                     1.0                  0.0  
1987                   0.0                     1.0                  0.0  
1988                   0.0                     1.0                  0.0  
1989                   0.0                     1.0                  0.0  
1990                   0.0                     1.0                  0.0  

[5 rows x 71 columns]

### Basic Statistics

### Plotting

### DiD and Regression